docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,298 @@
+
+- Some resource creations (say, `pods` or `LoadBalancer` `services`) could be throttled/rate-limited.
+- Even the request rate to the control-plane could be throttled/rate-limited in an intelligent way to so that critical control-plane and system components get the throughput they need but other non-critical components get throttled/rate-limited.
+- Gardener can defer any disruptive updates until later or ramp them down.
+
+## Alternatives
+
+### Gardener Shoot resource
+
+The `status.conditions` part of the Gardener `Shoot` [resource](#gardener-shoot-resource) could be used to publish and take action on the cluster health conditions.
+
+One possible disadvantage to this approach is that the the `Shoot` resource is hosted in the garden cluster, so any extension or controller that wants to take corrective or mitigating action based the cluster health conditions will need access to the garden cluster. Only, the `gardenlet` needs access to the garden cluster and it might be an anti-pattern to proliferate such access to other controllers in the seed cluster.
+
+A second disadvantage is that the `Shoot` resource is specific to the Gardener and some generic extensions or controllers might prefer to avoid adding a direct dependency to the `Gardener` project.
+
+### Gardener Extensions Cluster Resource
+
+One instance of the Gardener extensions `Cluster` [resource](#gardener-extensions-cluster-resource) is maintained per shoot cluster in the seed cluster whose control-plane is hosted in that seed cluster. So, in principle, this can be used for the purpose of publishing and consuming cluster health conditions.
+
+One of the disadvantages of this approach is that `Cluster` resource is [non-namespaced](https://github.com/gardener/gardener/blob/master/pkg/apis/extensions/v1alpha1/types_cluster.go#L26) which might be suitable for centrally deployed extensions and controllers deployed in the `garden` namespaces but may not be suitable for the shoot-specific controllers deployed per shoot control-plane (deployed in the shoot namespace in the seed cluster).
+
+A second disadvantage is that the part where the shoot status conditions can be maintained is [unstructured](https://github.com/gardener/gardener/blob/master/pkg/apis/extensions/v1alpha1/types_cluster.go#L57) making it unsuitable as an API for publishing and consuming actionable cluster health conditions.
+
+A third disadvantage is the same as the `Shoot` [resource](#gardener-shoot-resouce-1) approach, that this would require the extensions and controllers add a direct dependency to the Gardener project which may not always be desirable.
+
+### Custom Cluster Resource
+
+Instead of re-using [any](#proposal) [of](#gardener-shoot-resource) [the](#gardener-extensions-cluster-resource) available resources mentioned above, a new [`CustomResourceDefinition`](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#customresourcedefinitions) could be defined for this purpose.
```

Comment:
I didn't quite get the motivation for this from the following text. Also, I think that having a dedicated resource is unnecessary. I actually like the proposal of adding `.status.conditions[]` to the `extensions.gardener.cloud/v1alpha1.Cluster` resource.

Reply:
I agree. But see here for a [caveat](https://github.com/gardener/gardener/pull/1797#discussion_r371583319).